### PR TITLE
Performance Improvements

### PR DESCRIPTION
**Pull request by ikatza**
Changes to reduce the CPU time necessary to simulate the geant4 stage. I believe the individual  commits in this PR are self explanatory, and the changes transparent. 

There's also some styling and minor refactoring.

I've run a few tests on `sbndcode`, these changes half the necessary CPU time. No serious issue has been risen, but this PR definitively needs a full suit of tests.

### CODE REVIEW

**Comment by lgarren** [comment]
trigger build

The code still needs to be reviewed, but running the CI anyway to provide feedback.

**Comment by ikatza** [comment]
They seem to be failing because of some version mismatch:
```
145: ERROR: Version conflict -- dependency tree requires versions conflicting with current setup of product larsoft: version v09_48_00 vs v09_48_01
```


The SBND and ICARUS builds fail because they need https://github.com/SBNSoftware/sbncode/pull/261

The tests from that PR have plenty of warnings, for the most part I don't understand why these warnings are showing.

**Comment by lgarren** [comment]
@ikatza apologies, the experiments have not yet caught up to the latest larsoft release.  Will try again next week.

File: larsim/PhotonPropagation/SemiAnalyticalModel.cxx [hunk 1]
```diff
@@ -99,37 +90,37 @@ void SemiAnalyticalModel::Initialization()
     x_v.push_back(elem.first);
     y_v.push_back(elem.second);
   }
-  fL_abs_vuv = std::round(interpolate(x_v, y_v, 9.7, false)); // 9.7 eV: peak of VUV emission spectrum   // TO DO UNHARDCODE FOR XENON 
+  fvuv_absorption_lenght = std::round(interpolate(x_v, y_v, 9.7, false)); // 9.7 eV: peak of VUV emission spectrum   // TO DO UNHARDCODE FOR XENON
```
Comment (knoepfel) [inline larsim/PhotonPropagation/SemiAnalyticalModel.cxx]: Suggest re-spelling `lenght` as `length` here and below.

File: larsim/PhotonPropagation/SemiAnalyticalModel.cxx [hunk 2]
```diff
@@ -239,12 +231,12 @@ SemiAnalyticalModel::VUVVisibility(geo::Point_t const& ScintPoint, OpticalDetect
...
   else {
     throw cet::exception("SemiAnalyticalModel")
-        << "Error: Invalid optical detector shape requested - configuration error in semi-analytical model, only rectangular, dome or disk optical detectors are supported." << "\n";
+      << "Error: Invalid optical detector shape requested - configuration error in semi-analytical model, only rectangular, dome or disk optical detectors are supported." << "\n";
   }
 
   // calculate visibility by geometric acceptance
   // accounting for solid angle and LAr absorbtion length
-  double visibility_geo = std::exp(-1. * distance / fL_abs_vuv) * (solid_angle / (4 * CLHEP::pi));
+  double visibility_geo = std::exp(-1. * distance / fvuv_absorption_lenght) * (solid_angle / (4 * CLHEP::pi));
```
Comment (knoepfel) [inline larsim/PhotonPropagation/SemiAnalyticalModel.cxx]: We understand that you did not change the logic here (just the name of the variable).  But is there any possibility of dividing by zero?  Or does the interpolation done above ensure that that can't happen?

**Review by knoepfel — COMMENTED** [review]
Changes look reasonable, @ikatza.  Only a couple very minor comments/questions.

**Comment by lgarren** [comment]
trigger build

File: larsim/PhotonPropagation/SemiAnalyticalModel.cxx (hunk 1 above)
Comment (ikatza) [reply larsim/PhotonPropagation/SemiAnalyticalModel.cxx]: Ups! fixed now

File: larsim/PhotonPropagation/SemiAnalyticalModel.cxx (hunk 2 above)
Comment (ikatza) [reply larsim/PhotonPropagation/SemiAnalyticalModel.cxx]: There is some possibility, depending on the input parameters. I've encapsulated the calculation on a function, check that is not 0 or less and made the variable a `const`.

**Comment by lgarren** [comment]
@ikatza See the icarus and sbnd build failures.  Looks like a matching PR will be needed for them.

**Comment by ikatza** [comment]
Hi @lgarren, there is one for both: https://github.com/SBNSoftware/sbncode/pull/261

In the thread there compilation tests are satisfied, and I believe the warnings are mostly unrelated. 

I'm looking at the output of the DUNE tests and it seems there are plenty of warnings for ProtoDUNE, from small variations of products sizes. Not sure if these are meaningful or related to this PR.

DUNE FD seems to be fine.

**Review by pgreen135 — COMMENTED** [review]
@ikatza is this attempting to pre-generate all possible transport time parameterisations? This was intentionally avoided as in many cases only a subset are ever required. The overhead from checking the number of dimension is negligible, so I would expect this to be _much_ slower in many circumstances

**Comment by pgreen135** [comment]
@ikatza can you also run the SBND validation suite, and post the results here (or in the SBN PR)

**Comment by pgreen135** [comment]
@lgarren could you not merge this until these checks are complete, as this could significantly impact the simulation in both SBND and DUNE if there is a problem...

**Comment by ikatza** [comment]
Hi @pgreen135 

Yes, it pre-generates all parameters. Overhead is compensated if the job has more than a handful of events and/or they're fairly busy events. I suspect that only jobs that generate single MIPs with less than 20 events would be slower. As a reference, it did pass the SBND CI quick single test and that only has two events.

I agree that I shouldn't be merged hastily, and that a full suite of tests are in order.

**Comment by lgarren** [comment]
@pgreen135 agreed

@ikatza will run CI tests again with SBNSoftware/sbncode#261

**Comment by lgarren** [comment]
trigger build with pull requests #90, SBNSoftware/sbncode#261

**Comment by ikatza** [comment]
I mistakenly swapped the logic of the `binary_search` algorithm in `SimDriftElectrons_module`. That was quite an omission from my part.

Annoyingly SBND test wasn't sensible to my mistake, fortunately ICARUS was.

On the flip side, it doesn't seem there's any issue related to the light component.

**Comment by pgreen135** [comment]
@ikatza I've run some tests with/without the timing pre-generation commit, and this does seem to be faster this way in most circumstances. The other changes in the light simulation seem sensible, so I am happy with this from my side. I would still suggest running the SBND validation suite, now that we have it -- see https://sbnsoftware.github.io/sbn/sbnci_wiki/sbnci_main for instructions

**Comment by lgarren** [comment]
@pgreen135 I'm about to trigger the regular CI here.  Are you able to run the validation with this PR and SBNSoftware/sbncode#261?  Running validation suites is generally the responsibility of the experiment.

**Comment by lgarren** [comment]
trigger build with pull requests #90, SBNSoftware/sbncode#261

**Comment by pgreen135** [comment]
@lgarren yes Iker is currently running the validation (had some technical difficulties that Vito is helping with, hence the delay). We will report the results here once its done

**Comment by lgarren** [comment]
Thanks @pgreen135

**Comment by lgarren** [comment]
FYI: I am able to make a local build against the MicroBooNE code.  There are unit test failures that are unrelated to this PR. (MicroBooNE is not yet up to date with  the latest larsoft and wirecell 0.18.0)

**Comment by ikatza** [comment]
We're still looking into validation, there has been some more technical difficulties along the way.

**Comment by lgarren** [comment]
Thanks for the update @ikatza

**Comment by ikatza** [comment]
We have now managed to run the validation tests on SBND. However, I'm just learning that the current CI system is incapable of testing changes that happen at the `g4` stage. People are working to include that capability.

Given the potential impact of these changes, I believe the best is to wait until I'm able to run the full validation test. It's looking like the tests will be ready by early next week.

**Comment by lgarren** [comment]
@ikatza No problem.  It's already too late for this week.  ;)

**Comment by knoepfel** [comment]
@ikatza, is there any update on this PR?

**Comment by ikatza** [comment]
The SBN CI system needed to be overhauled to be able to test the g4 stage. The overhauling is in principle finished, but we keep getting errors from various systems. We're still working on it.

I'll update this thread as soon as we have some results.

**Comment by ikatza** [comment]
I finally have some validation tests to show. 

Since the last time I commented in this thread, I pushed some more changes to improve performance in other places. I have tried these changes for DUNE simulations and in some cases these improvements shorten the spent time to a third. 

Importantly, commit [b521f2c](https://github.com/LArSoft/larsim/pull/90/commits/b521f2cef6782cafb8ee905fe91a4fcb3ed4f892) prevents the missing scenarios in which channels with no detected photos were being added to the `OpDetBacktrackerRecord`. I do not know if this omission was the cause for any issues with the PhotonBackTracker, but it does fix the code to match the excepted behaviour. 

Remember that tests need to pick up PR https://github.com/SBNSoftware/sbncode/pull/261, which I have updated to match this version. 

The validation results:

TPC Simulation
https://dbweb9.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/9615&platform=Linux%203.10.0-1160.59.1.el7.x86_64&phase=ci_validation&buildtype=slf7%20e20:prof

PDS Simulation
https://dbweb9.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/9602&platform=Linux%203.10.0-1160.59.1.el7.x86_64&phase=ci_validation&buildtype=slf7%20e20:prof

**Comment by pgreen135** [comment]
Thanks @ikatza for running the validation. From the SBND side I am happy with these changes.

**Comment by lgarren** [comment]
trigger build

**Comment by ikatza** [comment]
SBND and ICARUS failed because the build didn't include https://github.com/SBNSoftware/sbncode/pull/261

My bad sorry, I didn't link it above. I have edited the comment now.

**Comment by lgarren** [comment]
trigger build with pull requests #90, SBNSoftware/sbncode#261

**Comment by knoepfel** [comment]
approve

**Pull request merged**